### PR TITLE
Better size estimation

### LINKED ISSUE #231: The size estimation is wrong for multi-container apps if the images share some layers
Connects-to: https://github.com/balena-io/balena-cli/issues/2043

### PR DESCRIPTION
**Pull request by zvin**
Fixes #231

### CODE REVIEW

**Comment by dfunckt** [comment]
I'm not sure about this change @zvin, it isn't fully addressing the linked issue (results are pretty much arbitrary) and at the same time gives worse results for the most common use case.

This is because the API stores the size of the image on disk (so the old code was accurate for the common use case), whereas the registry will reply with the amount of bytes that will be downloaded. Apparently you need the former, and the only way to get an accurate result without counting shared layers multiple times is to actually pull the images and query the layer sizes on disk via the local daemon.

File: package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "get-port": "^3.2.0",
     "lodash": "^4.17.20",
     "node-cleanup": "^2.1.2",
+    "request-promise": "^4.2.6",
```
[the remark this answers is not in the document]
Comment (zvin) [reply package.json:32]: I know, but it (request) is already installed by dependencies.
We can use axios instead with very few changes.

**Comment by zvin** [comment]
@dfunckt why do you think the results are arbitrary ?
We sum the sizes of `tar`ed layers. It will be slightly larger (around 3% in my tests) than the size of the layers on the disk.

Today, the common use case is multi-container apps. And most multi-container apps share some layers.
Here https://jel.ly.fish/847be677-7404-4c60-8fa6-b6344b4f4ae8 the user complains that we create a 19GiB image for a device with 16GiB storage.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (zvin) [reply package.json:32]: This is what was used before `v6.0.0`, before we had the image size in the api.

**Comment by pdcastro** [comment]
> the user complains that we create a 19GiB image for a device with 16GiB storage.

Shaun summarised it in this CLI issue: https://github.com/balena-io/balena-cli/issues/2043

**Comment by dfunckt** [comment]
Results are just an estimate @zvin based on your heuristic that tar'd layers end up 3% larger. Can't we retrieve absolute numbers? What's wrong with pulling the images and doing the calculation locally? You *are* going to pull the images later on, right?

Also, I think the problem where we may end up with an image that is larger than what would fit on the device is orthogonal and still has to be handled regardless of how this PR ends up.

**Comment by zvin** [comment]
>What's wrong with pulling the images and doing the calculation locally? 
We need a place to pull them into.

**Comment by dfunckt** [comment]
I saw you mentioned resize2fs in FD -- would creating space large enough to hold the images as before and then calculating the actual size and resizing it down work? Otherwise, can you not create temporary space and copy the layers over when done?

**Comment by dfunckt** [comment]
@pdcastro [described it really nicely](https://github.com/balena-io/balena-cli/issues/2043#issuecomment-697030561) in that issue thread:

> Perhaps a solution might be to add a "shrinking" step after successful preloading, reducing the amount of free disk space in the image file. This would require "disk defragmenting", but I think there are some Linux tool that can achieve it.

**Comment by zvin** [comment]
>would creating space large enough to hold the images as before and then calculating the actual size and resizing it down work? 

It would, this PR still stands though. We need to estimate the required space for pulling the images.
Imagine an application with 100 slightly different containers sharing a 2GiB layer. With this PR we will allocate ~2GiB, before this PR we would allocate 200GiB.

>Otherwise, can you not create temporary space and copy the layers over when done?
I guess we could but that brings other problems, for example if some docker images are already preloaded on that disk image.

**Comment by zvin** [comment]
>Perhaps a solution might be to add a "shrinking" step after successful preloading

Yes, but we still need a better estimation of the required space.

**Comment by dfunckt** [comment]
> It would, this PR still stands though.

Okay, as long as you pad the estimate generously enough (currently 40%, right?) so we don't run out of space and fail, I'm with you.

**Comment by pdcastro** [comment]
> [...] can you not create temporary space and copy the layers over when done?
>> I guess we could but that brings other problems, for example if some docker images are already preloaded on that disk image.

What about cloning the original image file as a "draft image", enlarging it with oversized estimations, pulling the images there, computing how much space was actually needed, then deleting the image and preloading the original one with the right size?

Probably no better than `resize2fs`. :-)

**Comment by zvin** [comment]
>What about cloning the original image file as a "draft image", enlarging it with oversized estimations, pulling the images there, computing how much space was actually needed, then deleting the image and preloading the original one with the right size?

I think shrinking after preload would be simpler and faster (no copies).

@shaunmulligan do we still need to support `btrfs` ? Is anyone still preloading 1.x images ?

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (Page-) [reply package.json:32]: @pdcastro it's deprecated insofar as it won't have any new features or breaking changes, it's still fine to use but it won't ever have native promise support or any other modern js stuff

**Comment by pdcastro** [comment]
@zvin, just checking, are you waiting from Shaun's answer re `btrfs` support and/or intending to make further changes to this PR before it can be approved and merged?

**Comment by zvin** [comment]
@pdcastro No, Im' asking about `btrfs` for shrinking after preloading which should be in a separate pull request.

**Review by pdcastro — APPROVED** [review]

**Pull request merged**